src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -26,65 +26,60 @@ public sealed class CompositeActionHandler : Handler, ICompositeActionHandler
-                step.ExecutionContext.ExpressionValues["github"] = gitHubContext;
+                // Inputs of the composite step
+                var inputsData = new DictionaryContextData();
+                foreach (var i in Inputs)
+                {
+                    inputsData[i.Key] = new StringContextData(i.Value);
+                }
 
-                // Set GITHUB_ACTION_PATH
-                step.ExecutionContext.SetGitHubContext("action_path", ActionDirectory);
+                // Temporary hack until after M271-ish. After M271-ish the server will never send an empty
+                // context name. Generated context names start with "__"
+                var childScopeName = ExecutionContext.GetFullyQualifiedContextName();
+                if (string.IsNullOrEmpty(childScopeName))
+                {
+                    childScopeName = $"__{Guid.NewGuid()}";
+                }
 
-                compositeSteps.Add(step);
-            }
+                // Create nested steps
+                var nestedSteps = new List<IStep>();
+                foreach (Pipelines.ActionStep stepData in Data.Steps)
+                {
+                    var step = HostContext.CreateService<IActionRunner>();
+                    step.Action = stepData;
+                    step.Stage = stage;
+                    step.Condition = stepData.Condition;
+                    step.ExecutionContext = ExecutionContext.CreateEmbeddedChild(childScopeName, stepData.ContextName);
+                    step.ExecutionContext.ExpressionValues["inputs"] = inputsData;
+                    step.ExecutionContext.ExpressionValues["steps"] = ExecutionContext.Global.StepsContext.GetScope(childScopeName);
+
+                    // Shallow copy github context
+                    var gitHubContext = step.ExecutionContext.ExpressionValues["github"] as GitHubContext;
+                    ArgUtil.NotNull(gitHubContext, nameof(gitHubContext));
+                    gitHubContext = gitHubContext.ShallowCopy();
+                    step.ExecutionContext.ExpressionValues["github"] = gitHubContext;
+
+                    // Set GITHUB_ACTION_PATH
+                    step.ExecutionContext.SetGitHubContext("action_path", ActionDirectory);
+
+                    nestedSteps.Add(step);
+                }
 
-            try
-            {
-                // This is where we run each step.
-                await RunStepsAsync(compositeSteps);
+                // Run nested steps
+                await RunStepsAsync(nestedSteps);
 
-                // Get the pointer of the correct "steps" object and pass it to the ExecutionContext so that we can process the outputs correctly
+                // Set outputs
                 ExecutionContext.ExpressionValues["inputs"] = inputsData;
-                ExecutionContext.ExpressionValues["steps"] = ExecutionContext.Global.StepsContext.GetScope(ExecutionContext.GetFullyQualifiedContextName());
```

Comment:
nothing changed here, just re-using the local variable instead since it's more descriptive (better for readability)